jax/_src/lax/lax.py

```diff
@@ -4534,23 +4534,35 @@ def _dilate_shape(shape, dilation):
 def _ceil_divide(x1, x2):
   return -np.floor_divide(np.negative(x1), x2)
 
+
+jax_enum_names = [*[es.name for es in xla_client.PaddingType], 'PYTORCH_SAME']
+PaddingType = enum.Enum(
+    'JaxPaddingType',
+    jax_enum_names,
+    start=0,
+)
+
 def padtype_to_pads(in_shape, window_shape, window_strides, padding):
   """Convert padding string to list of pairs of pad values."""
-  PaddingType = xla_client.PaddingType
-
   if isinstance(padding, str):
-    mapping = {'VALID': PaddingType.VALID, 'SAME': PaddingType.SAME}
     try:
-      padding = mapping[padding.upper()]
+      padding = PaddingType[padding.upper()]
     except KeyError as err:
       msg = "Unrecognized padding type: expected 'VALID' or 'SAME', got {}."
       raise RuntimeError(msg.format(padding)) from err
 
-  if padding == PaddingType.SAME:
+  if padding in [PaddingType.SAME, PaddingType.PYTORCH_SAME]:
     out_shape = _ceil_divide(in_shape, window_strides)
-    pad_sizes = np.maximum(0, (out_shape - 1) * window_strides +
-                                window_shape - in_shape)
-    return [(pad_size // 2, pad_size - pad_size // 2) for pad_size in pad_sizes]
+    pad_sizes = np.maximum(
+        0,
+        (out_shape - 1) * window_strides + window_shape - in_shape,
+    )
+    if padding == PaddingType.PYTORCH_SAME:
+      shift_pad_sizes = pad_sizes + 1
```

Comment:
Also, I can't prove to myself that this method is actually equivalent to what PyTorch is doing. The PyTorch code you linked does roughly this:
```
pad_size = ...
left_pad = pad_size // 2
right_pad = pad_size - left_pad
```
So, if `pad_size` is odd, the excess padding will be put on the right. But your code here favors moving it to the left, because of the `+ 1`?

Reply:
Yeah, it doesn't because PyTorch code uses a bit different definition. However, to make sure that it works right, I added a correspondent test that contains a copy of padding from PyTorch.

Reply:
> Yeah, it doesn't because PyTorch code uses a bit different definitions. However, to make sure that it works right, I added a correspondent tests that contains a copy of padding from PyTorch.

Sorry, I don't follow. Can you please elaborate on why this method is equivalent?

Reply:
Okay, let's consider Haiku implementation first:
```
def same(effective_kernel_size: int) -> Tuple[int, int]:
  """Pads such that the output size matches input size for stride=1."""
  return ((effective_kernel_size - 1) // 2, effective_kernel_size // 2)
```
Here ```same``` operates with ```effective_kernel_size``` which is calculated as ```effective_kernel_size=(kernel_size-1)*rate + 1``` [[Link](https://github.com/deepmind/dm-haiku/blob/main/haiku/_src/pad.py#L86)]

Now let's look at PyTorch:
```
for d, k, i in zip(dilation, kernel_size,
                                   range(len(kernel_size) - 1, -1, -1)):
  total_padding = d * (k - 1)
  left_pad = total_padding // 2
  self._reversed_padding_repeated_twice[2 * i] = left_pad
  self._reversed_padding_repeated_twice[2 * i + 1] = (total_padding - left_pad)
```

So here ```total_padding=effective_kernel_size-1```.  ```total_padding - left_pad ``` is equivalent to ```effective_kernel_size // 2``` because ```total_padding = left_p - right_p = 0.5 left_p - right_p -> right_p = 0.5 total_padding```.

I hope this helps.